### PR TITLE
Prevent crashing on empty body requests

### LINKED ISSUE #7349: Crash in staging project selection
unstaging a request without body is not validated and causes a crash:

```
I, [2019-04-05T18:11:12.207140 #13395]  INFO -- : [e76c17ce-3c18-44a7-b72b-d7960720c3de] [13395:12350.80] method=DELETE path=/staging/openSUSE:Factory/staging_projects/openSUSE:Factory:Staging:C/staged_requests format=xml 
controller=Staging::StagedRequestsController action=destroy status=500 error='NoMethodError: undefined method `elements' for {}:Hash' duration=115.02 view=0.00 db=29.28 params={"staging_workflow_project"=>"openSUSE:Factory
", "staging_project_name"=>"openSUSE:Factory:Staging:C"} host=149.44.161.60 time=435536.26 backend=0 user=coolo
F, [2019-04-05T18:11:12.214487 #13395] FATAL -- : [e76c17ce-3c18-44a7-b72b-d7960720c3de] [13395:12350.80]   
F, [2019-04-05T18:11:12.214569 #13395] FATAL -- : [e76c17ce-3c18-44a7-b72b-d7960720c3de] [13395:12350.80] NoMethodError (undefined method `elements' for {}:Hash):
F, [2019-04-05T18:11:12.214603 #13395] FATAL -- : [e76c17ce-3c18-44a7-b72b-d7960720c3de] [13395:12350.80]   
F, [2019-04-05T18:11:12.214681 #13395] FATAL -- : [e76c17ce-3c18-44a7-b72b-d7960720c3de] [13395:12350.80] app/controllers/staging/staged_requests_controller.rb:57:in `request_numbers'
[e76c17ce-3c18-44a7-b72b-d7960720c3de] app/controllers/staging/staged_requests_controller.rb:37:in `destroy'
[e76c17ce-3c18-44a7-b72b-d7960720c3de] config/initializers/wrap_parameters.rb:38:in `call'

### PR DESCRIPTION
**Pull request by eduardoj**
Also error messages are thrown in case of a XML request with:
- empty body
- invalid XML
- no parseable requests' numbers

Fixes #7349.

Co-authored-by: David Kang <dkang@suse.com>

### CODE REVIEW

**Comment by eduardoj** [comment]
[Address Feedback](https://trello.com/c/6UXRNR8X/1294-address-feedback)

File: src/api/app/controllers/staging/staged_requests_controller.rb [hunk 1]
```diff
@@ -53,12 +53,28 @@ def destroy
...
+      errorcode: 'invalid_request',
+      message: 'Error while parsing the numbers of the requests'
+    )
   end
 
-  def xml_hash
-    Xmlhash.parse(request.body.read) || {}
+  def set_xml_hash
+    request_body = request.body.read
+    @parsed_xml = Xmlhash.parse(request_body) if request_body.present?
```
Comment (vpereira) [inline src/api/app/controllers/staging/staged_requests_controller.rb]: I think we could make it a global validator used by all `update` actions on the API, what do you think @eduardoj ?

Comment (eduardoj) [reply src/api/app/controllers/staging/staged_requests_controller.rb]: I think the same. Everything that requires an xml should use validation of empty body and xml validation.

Comment (hennevogel) [reply src/api/app/controllers/staging/staged_requests_controller.rb]: We already have that: [ApplicationController.validate_xml_request](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/controllers/application_controller.rb#L380)

File: src/api/app/controllers/staging/staged_requests_controller.rb [hunk 2]
```diff
@@ -53,12 +53,28 @@ def destroy
...
+    return if @parsed_xml
+
+    error_options = if request_body.present?
+                      { status: 400, errorcode: 'invalid_xml_format', message: 'XML format is not valid' }
```
Comment (hennevogel) [inline src/api/app/controllers/staging/staged_requests_controller.rb]: What is not valid and what can I do about this?

Comment (eduardoj) [reply src/api/app/controllers/staging/staged_requests_controller.rb]: Being the mse
@hennevogel:
> What is not valid and what can I do about this?

Being the message "XML format is not valid", what is not valid is the XML format. What can one do about this: fix the XML format.

Do you think the message is not clear enough? What do you propose instead?

File: src/api/app/controllers/staging/staged_requests_controller.rb (hunk 1 above)
Comment (eduardoj) [reply src/api/app/controllers/staging/staged_requests_controller.rb]: Before I meant that everything that requires an xml should use validation of empty body and well-formed xml.

Method `validate_xml_request` validates the given text against a xml schema definition file (.xsd files). Maybe validation through xml schemas is too much at this point.

File: src/api/app/controllers/staging/staged_requests_controller.rb (hunk 2 above)
Comment (hennevogel) [reply src/api/app/controllers/staging/staged_requests_controller.rb]: Hm, I guess real validation with `Suse::Validator` but you are right, not now.

File: src/api/app/controllers/staging/staged_requests_controller.rb (hunk 1 above)
Comment (hennevogel) [reply src/api/app/controllers/staging/staged_requests_controller.rb]: It is indeed. All of this is still in flux.

**Comment by hennevogel** [comment]
codecov hangs. coveralls is fine. merging!

**Pull request merged**